tachyon/math/finite_fields/quartic_extension_field.h

```diff
@@ -0,0 +1,662 @@
+    //   = c0² + (c2² + 2 * c1 * c3) * q + 2 * (c0 * c1 + c2 * c3 * q) * x + (c1² + 2 * c0 * c2 + c3 * q) * x² + 2 * (c0 * c3 + c1 * c2) * x³
+    //   = (c0² + (c2² + 2 * c1 * c3) * q, 2 * (c0 * c1 + c2 * c3 * q), c1² + 2 * c0 * c2 + c3 * q, 2 * (c0 * c3 + c1 * c2))
+    // where q is |Config::kNonResidue|.
+
+    // See https://eprint.iacr.org/2006/471.pdf
+    // Devegili OhEig Scott Dahab --- Multiplication and Squaring on AbstractPairing-Friendly Fields.pdf; Section 5
+    // clang-format on
+
+    constexpr BaseField kInv2 = *BaseField(2).Inverse();
+    constexpr BaseField kInv3 = *BaseField(3).Inverse();
+    constexpr BaseField kInv4 = *BaseField(4).Inverse();
+    constexpr BaseField kInv6 = *BaseField(6).Inverse();
+    constexpr BaseField kInv12 = *BaseField(12).Inverse();
+    constexpr BaseField kInv20 = *BaseField(20).Inverse();
+    constexpr BaseField kInv24 = *BaseField(24).Inverse();
+    constexpr BaseField kInv30 = *BaseField(30).Inverse();
+    constexpr BaseField kInv120 = *BaseField(120).Inverse();
+    constexpr BaseField kNeg5 = -BaseField(5);
+    constexpr BaseField kNegInv2 = -kInv2;
+    constexpr BaseField kNegInv3 = -kInv3;
+    constexpr BaseField kNegInv4 = -kInv4;
+    constexpr BaseField kNegInv6 = -kInv6;
+    constexpr BaseField kNegInv12 = -kInv12;
+    constexpr BaseField kNegInv24 = -kInv24;
+    constexpr BaseField kNegInv120 = -kInv120;
+
+    // h1 = 2 * c1
+    BaseField h1 = a.c1_.Double();
+    // h2 = 4 * c2
+    BaseField h2 = a.c2_.Double();
+    h2.DoubleInPlace();
+    // h3 = 8 * c3
+    BaseField h3 = a.c3_.Double();
+    h3.DoubleInPlace().DoubleInPlace();
+
+    // v0 = c0²
+    BaseField v0 = a.c0_.Square();
+    // v1 = (c0 + c1 + c2 + c3)²
+    BaseField v1 = (a.c0_ + a.c1_ + a.c2_ + a.c3_).Square();
+    // v2 = (c0 - c1 + c2 - c3)²
+    BaseField v2 = (a.c0_ - a.c1_ + a.c2_ - a.c3_).Square();
+    // v3 = (c0 + 2 * c1 + 4 * c2 + 8 * c3)²
+    BaseField v3 = (a.c0_ + h1 + h2 + h3).Square();
+    // v4 = (c0 - 2 * c1 + 4 * c2 - 8 * c3)²
+    BaseField v4 = (a.c0_ - h1 + h2 - h3).Square();
+    // h1 = 3 * c1
+    h1 += a.c1_;
+    // h2 = 9 * c2
+    h2.DoubleInPlace().AddInPlace(a.c2_);
+    // h3 = 27 * c3
+    h3 += a.c3_;
+    h3 += h3.Double();
+    // v5 = (c0 + 3 * c1 + 9 * c2 + 27 * c3)²
+    BaseField v5 = (a.c0_ + h1 + h2 + h3).Square();
+    // v6 = c3²
+    BaseField v6 = a.c3_.Square();
+
+    // v0_5 = 5 * v0
+    BaseField v0_5 = v0.Double();
+    v0_5.DoubleInPlace().AddInPlace(v0);
+    // v6_3 = 3 * v6
+    BaseField v6_3 = v6.Double();
+    v6_3 += v6;
+
+    // clang-format off
+    // b.c0 = v0 +
+    //        q * ((1 / 4) * v0 - (1 / 6) * (v1 + v2) + (1 / 24) * (v3 + v4) - 5 * v6)
+    b.c0_ = v0 +
+            Config::MulByNonResidue(kInv4 * v0 + kNegInv6 * (v1 + v2) + kInv24 * (v3 + v4) + kNeg5 * v6);
+    // b.c1 = -(1 / 3) * v0 + v1 - (1 / 2) * v2 + (1 / 20) * v4 + (1 / 30) * v5 - 12 * v6 +
```

Comment:
I think you missed out kNegInv4 * v3 here.

Reply:
Nice catch!

Reply:
@chokobole There is a typo. It's - (1 / 4) * v3, not - (1 / 3) * v3.